# Data Exploration and Cleaning

### Dataset Resource Information
- **Dataset Source**: Shopping Dataset
- **Resource Link**: https://www.kaggle.com/datasets/anvitkumar/shopping-dataset
- **Local File Used**: `Dataset/Combined_dataset.csv`

### Objectives
1. Load a CSV dataset into a Pandas DataFrame.
2. Explore data (head/tail, shape, columns, data types).
3. Handle missing values (identify, fill/drop).
4. Perform basic operations (filter rows, select columns).
5. Remove duplicates.
6. Create a derived column (total_amount = price * quantity).
7. Save the cleaned dataset as a new CSV file.

### 1. Load the CSV Dataset
This code snippet imports the `pandas` and `numpy` libraries, and loads the CSV dataset into a DataFrame.
- **Functions used**: `pd.read_csv()` to read the CSV file.

In [ ]:
import pandas as pd
import numpy as np

dataset_path = "Dataset/Combined_dataset.csv"

df = pd.read_csv(dataset_path)

print("Dataset loaded successfully!")

Dataset loaded successfully!


### 2. Explore the Dataset
This code snippet explores the structure, metadata, statistics, and dimensions of the loaded DataFrame.
- **Functions/Attributes used**: `.shape` (get rows/cols count), `.columns` (list column names), `.dtypes` (check data types), `.head()` (view first rows), `.tail()` (view last rows), `.info()` (view memory and null information), and `.describe()` (view statistical summaries).

In [ ]:
print(f"Dataset contains {df.shape[0]} rows and {df.shape[1]} columns.\n")

print("Column list:")
print(df.columns.tolist())
print("\nData Types:")
print(df.dtypes.head(10))

Dataset contains 1000 rows and 24 columns.

Column list:
['product_id', 'title', 'product_description', 'rating', 'ratings_count', 'initial_price', 'discount', 'final_price', 'currency', 'images', 'delivery_options', 'product_details', 'breadcrumbs', 'product_specifications', 'amount_of_stars', 'what_customers_said', 'seller_name', 'sizes', 'videos', 'seller_information', 'variations', 'best_offer', 'more_offers', 'category']

Data Types:
product_id               int64
title                   object
product_description     object
rating                 float64
ratings_count            int64
initial_price            int64
discount               float64
final_price             object
currency                object
images                  object
dtype: object


In [ ]:
df.head()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,...,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
0,8376765,Lino Perros,Women Navy Blue Solid Backpack,3.8,15,3995,58.0,"""₹3,995.00""",INR,http://assets.myntassets.com/assets/images/837...,...,"{""1_star"":2,""2_stars"":0,""3_stars"":3,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-8376765_cae700""]",NaN,[{}],{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,67,2899,35.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/913...,...,"{""1_star"":3,""2_stars"":4,""3_stars"":2,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-9136281_cae700""]",NaN,"[{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,226,2999,65.0,"""₹2,999.00""",INR,http://assets.myntassets.com/assets/images/176...,...,"{""1_star"":9,""2_stars"":5,""3_stars"":10,""4_stars""...",NaN,NaN,"[{""size"":""S""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,"[{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1052,1675,52.0,"""₹1,675.00""",INR,http://assets.myntassets.com/assets/images/137...,...,"{""1_star"":33,""2_stars"":21,""3_stars"":69,""4_star...",NaN,NaN,"[{""size"":""Onesize""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
4,13939916,MYTRIDENT,Men Blue Solid Bath Robe,4.7,12,2899,17.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/pro...,...,"{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""M"",""value"":""45.0in"",""value_name"":""Fr...",NaN,NaN,"[{},{},{},{},{},{},{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on IndusI...",bath-robe


In [ ]:
df.tail(3)

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,...,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
997,1708108,GIORDANO,Women Rose Gold-Toned Mirrored Analogue Watch ...,2.8,5,7950,50.0,"""₹7,950.00""",INR,http://assets.myntassets.com/assets/images/170...,...,"{""1_star"":2,""2_stars"":1,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]",NaN,NaN,"[{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
998,19280114,Fastrack,Men Black Brass Dial & Black Stainless Steel S...,4.3,17,3295,20.0,"""₹2,636.00""",INR,http://assets.myntassets.com/assets/images/192...,...,"{""1_star"":1,""2_stars"":2,""3_stars"":0,""4_stars"":...",NaN,Tru███m R███il,"[{""size"":""Onesize""}]",NaN,Sho███o. ███-10█████████enu███████████████████...,NaN,"{""applicable_on"":""Orders above Rs. 1499 (only ...","[{""offer_name"":""10% Instant Discount on IndusI...",watches
999,18601682,BOLDFIT,Purple Textured Yoga Mat,3.8,161,699,31.0,"""₹699.00""",INR,http://assets.myntassets.com/assets/images/186...,...,"{""1_star"":24,""2_stars"":8,""3_stars"":22,""4_stars...",NaN,NaN,"[{""size"":""Medium""}]",NaN,NaN,"[{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",yoga-mats


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   product_id              1000 non-null   int64  
 1   title                   1000 non-null   object 
 2   product_description     1000 non-null   object 
 3   rating                  1000 non-null   float64
 4   ratings_count           1000 non-null   int64  
 5   initial_price           1000 non-null   int64  
 6   discount                879 non-null    float64
 7   final_price             1000 non-null   object 
 8   currency                1000 non-null   object 
 9   images                  1000 non-null   object 
 10  delivery_options        1000 non-null   object 
 11  product_details         1000 non-null   object 
 12  breadcrumbs             1000 non-null   object 
 13  product_specifications  1000 non-null   object 
 14  amount_of_stars         1000 non-null   o

In [ ]:
df.describe()

,product_id,rating,ratings_count,initial_price,discount
count,1.000000e+03,1000.0000,1000.000000,1000.00000,879.000000
mean,1.713437e+07,3.6215,76.691000,2723.24100,53.503982
std,3.813766e+06,1.3744,241.114263,2408.69736,18.029201
min,5.868460e+05,0.0000,0.000000,249.00000,1.000000
25%,1.511501e+07,3.7000,7.000000,1399.00000,44.000000
50%,1.820890e+07,4.1000,17.000000,1999.00000,56.000000
75%,1.902737e+07,4.3000,58.000000,3299.00000,66.000000
max,2.274707e+07,5.0000,4441.000000,22199.00000,88.000000


### 3. Handle Missing Values
This code snippet scans the dataset to identify missing values and fills them with appropriate defaults.
- **Functions used**: `.isna().sum()` (detect missing values), and `.fillna()` (replace missing NaN values with specified defaults).

In [ ]:
missing_values = df.isna().sum()
print("Columns with missing values:")
print(missing_values[missing_values > 0])

Columns with missing values:
discount               121
what_customers_said    573
seller_name            301
videos                 781
seller_information     301
variations             562
dtype: int64


In [ ]:
#1. Discount (0%)
df['discount'] = df['discount'].fillna(0.0)

#2. Seller name ("Unknown")
df['seller_name'] = df['seller_name'].fillna('Unknown')
df['seller_information'] = df['seller_information'].fillna('Unknown')

print("Remaining missing values:")
print(df.isna().sum()[['discount', 'seller_name', 'seller_information']])

Remaining missing values:
discount              0
seller_name           0
seller_information    0
dtype: int64


### 4. Filter Rows and Select Columns
This code snippet cleans string formats in `final_price` (removing the Rupee symbol and commas) to convert it to a numeric type, adds a mock `quantity` column, and filters rows.
- **Functions/Methods used**: `.str.replace()` and `.str.strip()` (text cleaning), `pd.to_numeric()` (convert data type to numeric), `np.random.randint()` (generate random integers), and boolean indexing (`df[...]`) for filtering.

In [ ]:
cleaned_price = df['final_price'].astype(str).str.replace('₹', '', regex=False)
cleaned_price = cleaned_price.str.replace(',', '', regex=False)
cleaned_price = cleaned_price.str.strip('"')

df['price'] = pd.to_numeric(cleaned_price, errors='coerce')

print("Sample of cleaned price:")
print(df[['final_price', 'price']].head())

Sample of cleaned price:
   final_price   price
0  "₹3,995.00"  3995.0
1  "₹2,899.00"  2899.0
2  "₹2,999.00"  2999.0
3  "₹1,675.00"  1675.0
4  "₹2,899.00"  2899.0


In [ ]:
np.random.seed(42)
df['quantity'] = np.random.randint(1, 6, size=len(df))

print("Sample of synthetic quantity column:")
print(df[['product_id', 'title', 'price', 'quantity']].head())

Sample of synthetic quantity column:
   product_id           title   price  quantity
0     8376765     Lino Perros  3995.0         4
1     9136281  Tommy Hilfiger  2899.0         5
2    17633752           Lavie  2999.0         3
3     1376949          F Gear  1675.0         5
4    13939916       MYTRIDENT  2899.0         5


In [ ]:
high_value_items = df[(df['price'] > 1000) & (df['category'].isin(['backpacks', 'clothing-set', 'casual-shoes']))]
print(f"Found {len(high_value_items)} high value items in selected categories.")

selected_columns = df[['product_id', 'title', 'category', 'price', 'quantity']]
selected_columns.head()

Found 18 high value items in selected categories.


,product_id,title,category,price,quantity
0,8376765,Lino Perros,backpacks,3995.0,4
1,9136281,Tommy Hilfiger,backpacks,2899.0,5
2,17633752,Lavie,backpacks,2999.0,3
3,1376949,F Gear,backpacks,1675.0,5
4,13939916,MYTRIDENT,bath-robe,2899.0,5


### 5. Remove Duplicates
This code snippet duplicates rows to demonstrate duplicate detection and removal.
- **Functions used**: `pd.concat()` (concatenate dataframes), `.duplicated()` (detect duplicate rows), and `.drop_duplicates()` (remove duplicate rows).

In [ ]:
df_with_duplicates = pd.concat([df, df.head(5)], ignore_index=True)
print(f"DataFrame size with duplicates: {df_with_duplicates.shape}")

duplicate_rows = df_with_duplicates[df_with_duplicates.duplicated()]
print(f"Number of duplicate rows identified: {len(duplicate_rows)}")

df_cleaned = df_with_duplicates.drop_duplicates()
print(f"DataFrame size after removing duplicates: {df_cleaned.shape}")

DataFrame size with duplicates: (1005, 26)
Number of duplicate rows identified: 5
DataFrame size after removing duplicates: (1000, 26)


### 6. Create a Derived Column
This code snippet calculates a new column named `total_amount` as a vector multiplication of the `price` and `quantity` columns.
- **Operators used**: Vectorized multiplication `*` operation across Pandas Series.

In [ ]:
df_cleaned['total_amount'] = df_cleaned['price'] * df_cleaned['quantity']

print("Derived column 'total_amount' created successfully!")
df_cleaned[['product_id', 'title', 'price', 'quantity', 'total_amount']].head()

Derived column 'total_amount' created successfully!


C:\Users\kotha\AppData\Local\Temp\ipykernel_23068\1998488026.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['total_amount'] = df_cleaned['price'] * df_cleaned['quantity']


,product_id,title,price,quantity,total_amount
0,8376765,Lino Perros,3995.0,4,15980.0
1,9136281,Tommy Hilfiger,2899.0,5,14495.0
2,17633752,Lavie,2999.0,3,8997.0
3,1376949,F Gear,1675.0,5,8375.0
4,13939916,MYTRIDENT,2899.0,5,14495.0


### 7. Save the Cleaned Dataset
This code snippet saves the cleaned DataFrame to a new CSV file.
- **Functions used**: `.to_csv()` with `index=False` to save the file.

In [ ]:
output_file = "cleaned_shopping_dataset.csv"
df_cleaned.to_csv(output_file, index=False)

print(f"Cleaned dataset saved successfully to '{output_file}'!")

Cleaned dataset saved successfully to 'cleaned_shopping_dataset.csv'!


### Data Cleaning Summary Comparison

| Characteristic | Raw Dataset | Cleaned Dataset |
| :--- | :--- | :--- |
| **Number of Rows** | 1000 | 1000 |
| **Price Data Type** | string (e.g. "₹3,995.00") | float64 (e.g. 3995.0) |
| **Quantity Column** | None | Added (int, 1-5) |
| **Total Amount Column** | None | Added (float, price * quantity) |
| **Missing Values (discount)** | 121 (NaN) | 0 (filled with 0.0) |
| **Missing Values (seller)** | 301 (NaN) | 0 (filled with 'Unknown') |